### DISCUSSION TITLE
async system tasks documentation?

### DISCUSSION THREAD

**Discussion by Jiehong** [Netflix/conductor#2450] [General]
Hi,

SystemTasks can be async or not (default): https://github.com/Netflix/conductor/blob/38379b84d05f8134544dd8a3a0c91f7fd6ad73ef/core/src/main/java/com/netflix/conductor/core/execution/tasks/WorkflowSystemTask.java#L63

However, the way they are handled is not entirely clear to me, and I couldn't find any doc per say.

This is what I've got so far, but correct me if I'm wrong:

## Async = false, system tasks only

Conductor will decide if the task must start or not based on a previous action.
Ex: start a workflow -> first system task is created and `start` is called.
Ex: task M is updated by a worker -> task N is scheduled by conductor, and `start` is called.

## Async = true, system tasks only

Conductor defers things to a queue.

If the system task is the only one in the workflow:

1. workflow started
2. conductor put the system tasks in an internal queue (https://github.com/Netflix/conductor/blob/586dc3e577869d49ad93709ca528feb9c4db3d70/core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java#L1505)
3. Later (when?), depending on the status of the tasks, `start` or `execute` is called (https://github.com/Netflix/conductor/blob/586dc3e577869d49ad93709ca528feb9c4db3d70/core/src/main/java/com/netflix/conductor/core/execution/AsyncSystemTaskExecutor.java#L116)
4. if the task is not isAsyncComplete, it gets re `execute` after workflow.system.task.worker.callback.seconds.

Is that correct?

Also, is there a timeout each time a system task's `start` or `execute` method is called?

**Comment by kishorebanala** [comment]
@Jiehong Your understanding is correct. 

> Also, is there a timeout each time a system task's start or execute method is called?

There are no blanket timeouts for System task execution, but task specific timeouts still apply.

**Reply by Jiehong** [reply]
thanks!
